[1단계 - 로또 구현] 플린트(이주환) 미션 제출합니다.

jhan0121 opened the pull request:
안녕하세요 리뷰어님! 7기 백엔드 크루 플린트라고 합니다.

로또 출력을 제대로 하는지 테스트를 하고 싶은데 Random값으로 생성되는 로또 클래스를 어떻게 테스트 해야할지 고민이었습니다.
그래서 List.of()를 사용해서 우리가 값을 지정해주고 그 지정해준 값들이 출력하고자 하는 문자열들을 생성되는지를 확인하는 방식으로 가닥을 잡았습니다.
이 때, Lotto 클래스 안에서 create 메서드를 사용했기 때문에 생성자가
```java
Lotto(){
this. lottoNumber = create();
}
```
이런 생성자 안에서 create메서드를 이용해서 Lotto클래스를 생성하는 느낌이었기에 Lotto의 lottoNumber private 변수를 원하는 대로 설정하기가 어려웠습니다.
최종적으로 저희는 LottoCreator라는 클래스를  새롭게 만들고 create메서드(랜덤 로또 번호들을 생성하여 List에 담아 이것을 return하는)를  옮기기로 하였습니다.
그 이후 Lotto 생성자에 List를 삽입하도록
```java
Lotto lotto = new Lotto(List<Integer> a) 
```
이런 방식으로 재구성하였습니다.
이 때 저희는 create라는 객체 생성을 본 객체가 아닌 다른 클래스에서 하도록 하는 것이 올바른지 많은 고민들을 하였으나 명확하게 정답을 내리지 못하여 이렇게 상황을 전달드립니다.

추가로 getter 사용을 지양하다보니 Service 클래스에 사용될 수도 있는 메서드들이 모델로 넘어가는 경향이 크다고 생각이 듭니다. 이 경우 어떻게 처리하는 것이 좋을지 궁금합니다.

추가: https://prolog.techcourse.co.kr/studylogs/3822

jhan0121 committed:
docs(README): 기능명세서 작성

jhan0121 committed:
feat: InputView 구현

(페어 매칭 1차)
로또 가격 입력 및 검증 기능

jhan0121 committed:
feat: 클래스 설계

Lotto, Lottos, WinnerNumber 생성

jhan0121 committed:
feat: Lotto 생성 기능 구현

(페어 2차)

jhan0121 committed:
test: Lotto 생성 테스트 및 구매 갯수 계산 테스트

jhan0121 committed:
fix: 단일 로또 정렬 기능 추가

(페어 3차)

jhan0121 committed:
feat: Lottos 개수만큼 로또 생성 기능 구현

jhan0121 committed:
test: Lottos 개수만큼 생성 테스트 구현

jhan0121 committed:
feat: OutputView 구현

구입금액안내, 개수 확인 문자열 출력

jhan0121 committed:
fix: Lotto 내 랜덤 생성 로직 LottoCreator로 이동

(4차 페어)

jhan0121 committed:
test: 로또 번호 출력 확인 테스트 생성

jhan0121 committed:
feat: LottoController run 로직 구현

(페어 5차)

jhan0121 committed:
feat: InputView 당첨번호 및 보너스 번호 입력 기능 구현

jhan0121 committed:
feat: OutputView 당첨 번호 출력 기능 구현

jhan0121 committed:
fix: InputView Scanner 오류 수정

(5차 페어)

jhan0121 committed:
feat: LottoResult 구현

jhan0121 committed:
fix: Readme 수익률 계산 공식 수정

jhan0121 committed:
feat: OutputView 당첨 통계 결과 출력 기능 구현

jhan0121 committed:
feat: Lottos getter 추가

jhan0121 committed:
feat: LottoService 구매 갯수, 당첨률 계산 기능 구현

jhan0121 committed:
feat: LottoController run 로직 구현

jhan0121 committed:
feat: WinnerNumber 당첨여부 확인 로직 구현

jhan0121 committed:
test: LottoResultTest 테스트 구현

jhan0121 committed:
test: LottoServiceTest 로또 결과 테스트 구현

jhan0121 committed:
feat: Main 로직 구현

jhan0121 committed:
refactor: 메서드 분할 및 스캐너 이동

jhan0121 committed:
test: Input 유효성 검사 테스트 추가

yebink commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,12 @@
+
+        LottoService lottoService = new LottoService();
+        LottoController lottoController = new LottoController(lottoService);
+        lottoController.run();
```
Service와 Controller를 사용하셨군요!

플린트가 구현한 코드에서 LottoService와 LottoController의 역할은 무엇일까요?
두 클래스 없이도 구현할 수 있을까요?  🤔

jhan0121 replied:
LottoService는 가격(price)를 매개변수로 받아서 구매한 로또의 갯수와 수익률을 계산하는 역할을 수행하고 있고, LottoController는 View에서 입력 받은 사용자의 입력을 기반으로 로또 관련 domain 객체(Lotto, Lottos, WinnerNumber)의 데이터를 조작하여 View로 전달하는 역할이라고 생각합니다!

Service는 설계에 따라 없어도 구현하는데 문제는 없을 것으로 생각합니다. 왜냐하면 필요한 객체에 메시지로 필요한 데이터를 전달을 적절히 처리해준다면 service는 단순 데이터 이동만 일어날 것이기 때문입니다.

Controller의 경우에는 없이 구현하려면 domain 간의 메시지 전달을 담당하는 객체를 따로 만들어 구현해야 할 것으로 생각되는데, 저는 해당 역할을 하는 객체가 Controller라고 생각하고 있어 Controller를 사용하는 것이 맞다고 생각합니다.

yebink replied:
플린트의 의견에 동의해요.

다만, Controller와 Service를 사용하다보면 로직이 분산되면서 객체지향적 사고를 하기가 어려워져요.
가령 `LottoController#compareWinning` 를 모델로 이동시킬 수 있을까요? 만약 그렇다면 어디가 가장 적절할까요?
Controller, Service, DTO 등의 개념은 나중에 Spring 을 본격적으로 사용할 때 더 학습해보시고, 
그 전까지는 먼저 위와 같은 고민을 많이 해보는 걸 추천해요 😄

yebink commented on README.md:
```diff
@@ -1,3 +1,17 @@
+
+- 구매 금액이 숫자가 아닌 경우, IllegalArgumentException을 throw한다.
+- 1000원 단위가 아닌 경우, IllegalArgumentException을 throw한다.
+- 당첨 번호 및 보너스 볼에 중복 번호를 입력할 경우, IllegalArgumentException을 throw한다.
```
요구 기능과 예외 사항을 잘 입력하셨네요! 👍🙂

yebink commented on src/test/java/View/InputViewTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    void 로또_구매_입력_오류_테스트(){
```
1개의 테스트에 동일한 체크를 하는 assert문이 2개 들어있네요.

JUnit 의 `@NullAndEmptySource` 를 사용해보면 어떨까요?

yebink commented on src/main/java/Model/LottoCreator.java:
```diff
@@ -0,0 +1,35 @@
+        return lottoNumber;
+    }
+
+    private static void addRandomNumber(List<Integer> lottoNumber) {
+        int count = LOTTO_COUNT;
+        Random random = new Random();
+
+        while (count > 0) {
+            int randomNumber = random.nextInt(LOTTO_MAX_NUMBER) + 1;
+            if (lottoNumber.contains(randomNumber)) {
+                continue;
+            }
+            lottoNumber.add(randomNumber);
+            count--;
+        }
+    }
```
indent 를 1개로 줄여볼까요~?

yebink commented on src/main/java/Model/Lottos.java:
```diff
@@ -0,0 +1,27 @@
+
+public class Lottos {
+
+    private List<Lotto> lottos;
```
👍

yebink commented on src/test/java/Model/LottoResultTest.java:
```diff
@@ -0,0 +1,22 @@
+class LottoResultTest {
+
+    @Test
+    void 개수_및_보너스_당첨_여부에_따라_당첨_횟수를_갱신해야_한다() {
```
테스트마다 이름을 잘 지정하셨네요!  <br/>

`@DisplayName` 어노테이션으로 테스트에 이름을 명시적으로 지정할 수 있어요.
지금 미션에서 반영할 필요는 없고, 2단계부터 시도해보세요 🙂

yebink reviewed (changes requested):
안녕하세요 플린트~
1단계 리뷰를 맡게 된 엘리입니다.

로또 미션 구현하느라 고생하셨어요!
커밋도 잘 나뉘어져 있고, TDD로 구현하기 위해 노력한 흔적이 보이네요! 👏 


> 저희는 create라는 객체 생성을 본 객체가 아닌 다른 클래스에서 하도록 하는 것이 올바른지 많은 고민들을 하였으나 명확하게 정답을 내리지 못하여 이렇게 상황을 전달드립니다.

우선 LottoCreator 로 접근하신 것도 좋은 방법이에요.
다만, Lotto 클래스에도 public 생성자가 열려 있는데요!
이렇게 되면, 나중에 `로또를 생성`하고자 하는 요구사항이 추가될 때, 누군가는 Lotto 클래스를 사용할 수도 있고, 또 다른 누군가는 LottoCreator 클래스를 사용할 수도 있겠죠?
그렇다면 Lotto 클래스에 생성자를 여러 개 두는 방법은 어떨까요?


> getter 사용을 지양하다보니 Service 클래스에 사용될 수도 있는 메서드들이 모델로 넘어가는 경향이 크다고 생각이 듭니다. 이 경우 어떻게 처리하는 것이 좋을지 궁금합니다.

아주 자연스러운 현상입니다. `Service 클래스에 사용될 수도 있는 메서드` 에는 어떤게 있을까요? 서비스 클래스의 역할에 대해 고민해보면 좋을 것 같아요.


실패하는 테스트가 2개 있으니 확인 부탁드려요.
그리고 이해가 안되거나 모르는 게 있으면 언제든 DM 주세요! 🙂

jhan0121 committed:
fix: 로또 확인 테스트 로직 오류 수정

비교값이 정렬되지 않아 출력이 일치하지 않은 문제 수정

jhan0121 committed:
refactor: InputViewTest 중복 테스트 로직 개선

중복 로직에 ParameterizedTest를 적용

jhan0121 committed:
refactor: LottoCreator 제거 및 Lotto 로직 수정

- LottoCreator에 있던 method Lotto 클래스로 이동
- Lotto 생성자 수정(사용자가 지정한 List를 저장하는 생성자 추가)

yebink reviewed (approved):
안녕하세요 플린트,

리뷰 잘 반영해주셨네요!
코멘트 하나 남겼으니 확인해주세요.

2단계에서도 화이팅해보아요
머지하겠습니다 🙂
